prow/cmd/webhook-server/validation.go

```diff
@@ -16,8 +16,96 @@ limitations under the License.
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
 
-func serveValidate(http.ResponseWriter, *http.Request) {
+	"github.com/sirupsen/logrus"
 
+	"k8s.io/api/admission/v1beta1"
+	apiv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/kube"
+	"k8s.io/test-infra/prow/plank"
+)
+
+var agentsNotSupportingCluster = sets.NewString("jenkins")
+
+func (wa *webhookAgent) serveValidate(w http.ResponseWriter, r *http.Request) {
+    body, err := ioutil.ReadAll(r.Body) 
+    if err != nil {
+        logrus.WithError(err).Fatal("Unable to read request")
+    }
+    var prowJob v1.ProwJob
+    admissionReview := &v1beta1.AdmissionReview{}
+    err = json.Unmarshal(body, admissionReview)
+	if err != nil {
+        logrus.WithError(err).Fatal("Unable to unmarshal admission review request")
+    }
+    admissionRequest := admissionReview.Request
+    err = json.Unmarshal(admissionRequest.Object.Raw, &prowJob)
+    if err != nil {
+        logrus.WithError(err).Fatal("Unable to unmarshal admission review request")
+    }
+    var admissionResponse *v1beta1.AdmissionResponse
+    if admissionRequest.Operation == "CREATE" {
+        if err := validateProwJobClusterOnCreate(prowJob, wa.statuses); err != nil {
+            admissionResponse = createAdmissionResponse(admissionRequest.UID, false, "Denied", err.Error())
+        } else {
+            admissionResponse = createAdmissionResponse(admissionRequest.UID, true, "Allowed", "")
+        }
+    }
+    admissionReview.Response = admissionResponse
+    resp, err := json.Marshal(admissionReview)
+    if err != nil {
+        logrus.WithError(err).Fatal("Unable to marshal response")
+    }
+    if _, err := w.Write(resp); err != nil {
+		logrus.WithError(err).Fatal("Unable to write response: %v", err)
+		http.Error(w, fmt.Sprintf("could not write response: %v", err), http.StatusInternalServerError)
+	}
+}
+
+func validateProwJobClusterOnCreate(prowJob v1.ProwJob, statuses map[string]plank.ClusterStatus) (error) {
+    if  prowJob.Spec.Cluster!= "" && prowJob.Spec.Cluster != kube.DefaultClusterAlias && agentsNotSupportingCluster.Has(string(prowJob.Spec.Agent)) {
+		return fmt.Errorf("%s: cannot set cluster field if agent is %s", prowJob.Name, prowJob.Spec.Agent)
+	}
+	if statuses != nil {
+		status, ok := statuses[prowJob.Spec.Cluster]
+		if !ok {
+			return fmt.Errorf("job configuration for %q specifies unknown 'cluster' value %q", prowJob.Name, prowJob.Spec.Cluster)
+		}
+		if status != plank.ClusterStatusReachable {
+			logrus.Warnf("Job configuration for %q specifies cluster %q which cannot be reached from Plank. Status: %q", prowJob.Name, prowJob.Spec.Cluster, status)
+		}
+    }
+    return nil
 }
+
+func createAdmissionResponse(uid types.UID, allowed bool, message string, reason string) *v1beta1.AdmissionResponse {
+    var ar *v1beta1.AdmissionResponse
+    if allowed {
```

Comment:
You can simplify this by defining the &apiv1.Status first. Have an if statement to determine if you need to set Reason, and then define AdmissionRepsonse